adapter-base/src/main/java/org/eclipse/hono/adapter/auth/device/TenantServiceBasedX509Authentication.java

```diff
@@ -256,6 +256,8 @@ public TenantServiceBasedX509Authentication(
         if (tenant.isAutoProvisioningEnabled(issuerDn)) {
             try {
                 authInfo.put(CredentialsConstants.FIELD_CLIENT_CERT, deviceCert.getEncoded());
+                authInfo.put(CredentialsConstants.FIELD_AUTO_PROVISION_AS_GATEWAY,
+                        tenant.isAutoProvisioningAsGatewayEnabled(issuerDn));
```

Comment:
why does a protocol adapter need to know whether the device will be auto-provisioned as a gateway or as a normal device? Isn't that the business of the registry implementation to determine? Having the adapter include this flag also allows the adapter to simply put in any value it wants to, right? FMPOV the registry must read the value of this flag from the tenant configuration on record and then act accordingly.

Reply:
Currently the protocol adapters decide if a device need to be auto-provisoned or not as below . So why not to decide if that should be a device or a gateway.


https://github.com/eclipse/hono/blob/86d0f934e1a1a409383ed48d8d30504362622326/adapter-base/src/main/java/org/eclipse/hono/adapter/auth/device/TenantServiceBasedX509Authentication.java#L256-L259


Also in the device registry implementations we check if the certificate is available in the _clientContext_ or not in order to determine if auto-provisioning is enabled or not. We don't retrieve the corresponding tenant's CA from the database and validate it. In the above approach in this PR, the device registry need not search again the underlying database for the corresponding tenant's CA to determine if a device or a gateway is being auto-provisioned.

Reply:
Obviously. But that is a problem FMPOV as it lets the protocol adapter decide if a device may be auto provisioned or not. The reason why the protocol adapter needs to know if auto-provisioning is enabled at all for the CA is because it needs to determine if it needs to include the client cert in the get Credentials request. However, the decision whether a device can actually be auto provisioned (and as what, a *device* or *gateway*) must be made by the registry, not the adapter.

Reply:
Do you suggest that the device registry should make use of the tenant's CA entry instead of verifying if a certificate is available in the _clientContext_ to determine if auto-provisioning is enabled or not? And to verify if a device or a gateway is to be auto-provisioned in the same fashion.

Reply:
exactly